contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+can be determined from the pod object directly.
+To summarize, device selection at the kubelet can be done in one of the following
+two ways:
+* Select device at pod admission while applying predicates and change all api
+  interfaces that are required to pass selected device to container manager.
+* Create resource consumption state again at container manager and select device.
+
+None of the above approach seems cleaner than doing device selection at scheduler,
+which helps to retain cleaner api interfaces between packages.
+
+## Scheduler Changes
+Scheduler already listens and maintains state in the cache for any changes in
+node or pod objects. We will enhance the logic:
+1. To listen and maintain the state in cache for user created *Resource Class* objects.
+2. To look for device related details in node objects and maintain accounting for
+   devices as well.
+
+From the events perspective, handling for the following events will be added/updated:
+
+### Resource Class Creation
+1. Initialize and add resource class info into local cache
+2. Iterate over all existing nodes in cache to figure out if there are devices
+   on these nodes which are selectable by resource class. If found, update the
+   resource class availability status in local cache.
+3. Patch the status of resource class api object with availability state in local
+   cache
+
+### Resource Class Deletion
+Delete the resource class info from the cache.
+
+### Node Addition
+Scheduler already caches `NodeInfo`. Now additionally update device state:
+1. Check in the node status if any devices are present.
+2. For each device found, iterate over all existing resource classes in the cache
+   to find resource classes which can select this particular device. For all
+   such resource classes, update the availability state in the local cache.
+3. ResourceClass api object's status, `ResourceClassStatus` will be patched
+   as per the updated availability state in the cache.
+
+### Node Deletion
+If node has devices which are selectable by existing resource classes:
+1. Adjust resource class state in local cache.
+2. Update resource class status by patching api object.
+
+### Pod Creation
+1. Get the requested resource class name and quantity from pod spec.
+2. Select nodes by applying predicates according to requested quantity and Resource
+   class's state present in the cache.
+3. On the selected node, select a Device from the stored devices info in cache
+   after matching key,value from requested resource class.
+4. After device selection, update(decrease) 'Requested' for all the resource
+   classes which could select this device in the cache.
+5. Patch the resource class objects with new 'Requested' in the `ResourceClassStatus`.
+6. Add the pod reference in local DeviceToPod mapping structure in the cache.
+7. Patch the pod object with selected device annotation with prefix 'scheduler.alpha.kubernetes.io/resClass'
```

Comment:
Would like to know the opinion of some sig-scheduling folks on the best way to achieve that.
Sending a patch request during host selection for a pod doesn't seem to be the right model.

Maybe @bsalamat or @timothysc ?

Reply:
Sending a patch request directly here is not right.

But we can probably update the annotation of `assumedPod`, and update Pod api object during `bind()`, which is async.

@bsalamat Make sense?

Reply:
I generally do not like the idea of scheduler knowing about mapping of devices to resource classes. This does not fit well in scheduler's logic and does not seem to be something that scheduler should be aware of.
It is also not clear why scheduler needs to update the pod with the selected device annotation.

Reply:
@bsalamat The problem I can see here is: device plugin is not expected to have any scheduling logic (e.g. select devices) by design, and if scheduler also does not want to have it. The only place to select devices will be kubelet. But what this proposal is describing requires awareness of whole picture of all devices info in the cluster to make decision, this is not what one kubelet is capable of (it only has device info of this node). 

I am willing to pushing this design as it do help to enable GPU topology in Kubernetes a lot, and also, fixes blockers for other devices like FPGA. So would be great to know your ideas for scheduling.